libs/dal/src/repositories/base-repository.ts

```diff
@@ -101,6 +110,109 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
     }
   }
 
+  private async createCursorBasedOrStatement({
+    isSortDesc,
+    paginateField,
+    after,
+    queryOrStatements,
+  }: {
+    isSortDesc: boolean;
+    paginateField?: string;
+    after: string;
+    queryOrStatements?: object[];
+  }): Promise<FilterQuery<T_DBModel>[]> {
+    const afterItem = await this.MongooseModel.findOne({ _id: after });
+    if (!afterItem) {
+      throw new DalException('Invalid after id');
+    }
+
+    let cursorOrStatements: FilterQuery<T_DBModel>[] = [];
+    let enchancedCursorOrStatements: FilterQuery<T_DBModel>[] = [];
+    if (paginateField && afterItem[paginateField]) {
+      const paginatedFieldValue = afterItem[paginateField];
+      cursorOrStatements = [
+        { [paginateField]: isSortDesc ? { $lt: paginatedFieldValue } : { $gt: paginatedFieldValue } } as any,
+        { [paginateField]: { $eq: paginatedFieldValue }, _id: isSortDesc ? { $lt: after } : { $gt: after } },
+      ];
+      const firstStatement = (queryOrStatements ?? []).map((item) => ({
+        ...item,
+        ...cursorOrStatements[0],
+      }));
+      const secondStatement = (queryOrStatements ?? []).map((item) => ({
+        ...item,
+        ...cursorOrStatements[1],
+      }));
+      enchancedCursorOrStatements = [...firstStatement, ...secondStatement];
+    } else {
+      cursorOrStatements = [{ _id: isSortDesc ? { $lt: after } : { $gt: after } }];
+      const firstStatement = (queryOrStatements ?? []).map((item) => ({
+        ...item,
+        ...cursorOrStatements[0],
+      }));
+      enchancedCursorOrStatements = [...firstStatement];
+    }
+
+    return enchancedCursorOrStatements.length > 0 ? enchancedCursorOrStatements : cursorOrStatements;
+  }
+
+  async cursorPagination({
+    query,
+    limit,
+    after,
+    sort,
+    paginateField,
+    enhanceQuery,
+  }: {
+    query?: FilterQuery<T_DBModel> & T_Enforcement;
+    limit: number;
+    after: number | string;
+    sort?: any;
+    paginateField?: string;
+    enhanceQuery?: (query: QueryWithHelpers<Array<T_DBModel>, T_DBModel>) => any;
+  }): Promise<{ data: T_MappedEntity[]; hasMore: boolean }> {
+    const isAfterNumber = typeof after === 'number';
+    const sortKeys = Object.keys(sort ?? {});
+    const isSortDesc = sortKeys.length > 0 && sort[sortKeys[0]] === -1;
+
+    let findQueryBuilder = this.MongooseModel.find({
+      ...query,
+    });
+    if (!isAfterNumber) {
+      const orStatements = await this.createCursorBasedOrStatement({
+        isSortDesc,
+        paginateField,
+        after,
+        queryOrStatements: query?.$or,
+      });
+
+      findQueryBuilder = this.MongooseModel.find({
+        ...query,
+        $or: orStatements,
+      });
+    }
+
+    findQueryBuilder.sort(sort).limit(limit + 1);
```

Comment:
take a one more item from the `limit` - this allows us to understand if there are more items available for the next page